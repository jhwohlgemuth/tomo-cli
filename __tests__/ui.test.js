import {useTemporaryDirectory} from './tomo-test';
import Queue from 'p-queue';
import React from 'react';
import {render} from 'ink-testing-library';
import Tomo, {Warning, Task, TaskList, populateQueue} from '../src/ui';

const ARROW_DOWN = '\u001B[B';
const {assign} = Object;

describe('populateQueue function', () => {
    test('can run with defaults', () => {
        populateQueue();
    });
    test('can run tasks with options and dispatch task results', async () => {
        const condition = async () => true;
        const task = jest.fn();
        const options = {foo: 'bar'};
        const dispatch = jest.fn();
        const tasks = [
            {condition, task},
            {condition, task},
            {condition, task}
        ];
        const queue = new Queue({concurrency: tasks.length});
        await populateQueue({queue, tasks, dispatch, options});
        expect(task.mock.calls.length).toBe(tasks.length);
        expect(dispatch.mock.calls.length).toBe(tasks.length + 1);
        const [passedOptions] = [...new Set(task.mock.calls.map(val => val[0]))];
        expect(passedOptions).toEqual(assign(options, {isNotOffline: true}));
        expect(dispatch.mock.calls).toMatchSnapshot();
    });
    test('can only run tasks that pass condition', async () => {
        const task = jest.fn();
        const options = {foo: 'bar'};
        const dispatch = jest.fn();
        const tasks = [
            {condition: async () => true, task},
            {condition: async () => false, task},
            {condition: async () => true, task},
            {condition: async () => false, task}
        ];
        const queue = new Queue({concurrency: tasks.length});
        await populateQueue({queue, tasks, dispatch, options});
        const [passedOptions] = [...new Set(task.mock.calls.map(val => val[0]))];
        expect(passedOptions).toMatchSnapshot();
        expect(task.mock.calls.length).toBe(2);
        expect(dispatch.mock.calls).toMatchSnapshot();
    });
    test('can catch task errors', async () => {
        const task = jest.fn();
        const options = {foo: 'bar'};
        const dispatch = jest.fn();
        const tasks = [
            {condition: async () => true, task},
            {condition: async () => {throw new Error();}, task},
            {condition: async () => true, task},
            {condition: async () => false, task}
        ];
        const queue = new Queue({concurrency: tasks.length});
        await populateQueue({queue, tasks, dispatch, options});
        expect(task.mock.calls.length).toBe(2);
        expect(dispatch.mock.calls[2][0].type).toBe('error');
    });
});
describe('Warning', () => {
    test('can render', () => {
        const callback = jest.fn();
        const {lastFrame} = render(<Warning callback={callback}>Hello World</Warning>);
        expect(lastFrame()).toMatchSnapshot();
    });
});
describe('Task component', () => {
    test('can render (pending)', () => {
        const text = 'test task text';
        const {lastFrame} = render(<Task text={text} isPending={true}></Task>);
        expect(lastFrame()).toMatchSnapshot();
    });
    test('can render (completed)', () => {
        const text = 'test task text';
        const {lastFrame} = render(<Task text={text} isComplete={true}></Task>);
        expect(lastFrame()).toMatchSnapshot();
    });
    test('can render (skipped)', () => {
        const text = 'test task text';
        const {lastFrame} = render(<Task text={text} isComplete={true} isSkipped={true}></Task>);
        expect(lastFrame()).toMatchSnapshot();
    });
    test('can render (errored)', () => {
        const text = 'test task text';
        const {lastFrame} = render(<Task text={text} isErrored={true}></Task>);
        expect(lastFrame()).toMatchSnapshot();
    });
    test('can render with default text', () => {
        const {lastFrame} = render(<Task isComplete={true}></Task>);
        expect(lastFrame()).toMatchSnapshot();
    });
});
xdescribe('TaskList component', () => {
    let tempDirectory;
    const skipInstall = true;
    const [setTempDir, cleanupTempDir] = useTemporaryDirectory();
    const ORIGINAL_CONSOLE_ERROR = console.error;//eslint-disable-line no-console
    beforeAll(() => {
        console.error = jest.fn();//eslint-disable-line no-console
    });
    afterAll(() => {
        console.error = ORIGINAL_CONSOLE_ERROR;//eslint-disable-line no-console
    });
    beforeEach(async () => {
        tempDirectory = await setTempDir();
        process.chdir(tempDirectory);
    });
    afterEach(async () => {
        await cleanupTempDir();
    });
    test('can render', done => {
        const options = {skipInstall};
        const {lastFrame} = render(<TaskList command={'add'} terms={['eslint']} options={options} done={complete}></TaskList>);
        function complete() {
            expect(lastFrame()).toMatchSnapshot();
            done();
        }
    });
});
describe('tomo', () => {
    let tempDirectory;
    const skipInstall = true;
    const [setTempDir, cleanupTempDir] = useTemporaryDirectory();
    const ORIGINAL_CONSOLE_ERROR = console.error;//eslint-disable-line no-console
    beforeAll(() => {
        console.error = jest.fn();//eslint-disable-line no-console
    });
    afterAll(() => {
        console.error = ORIGINAL_CONSOLE_ERROR;//eslint-disable-line no-console
    });
    beforeEach(async () => {
        tempDirectory = await setTempDir();
        process.chdir(tempDirectory);
    });
    afterEach(async () => {
        await cleanupTempDir();
    });
    test('add', () => {
        const input = ['add'];
        const {lastFrame, stdin} = render(<Tomo input={input} flags={{skipInstall}}/>);
        expect(lastFrame()).toMatchSnapshot();
        stdin.write(ARROW_DOWN);
        expect(lastFrame()).toMatchSnapshot();
    });
    xtest('add eslint', done => {
        const input = ['add', 'eslint'];
        const flags = {skipInstall};
        const {lastFrame} = render(<Tomo input={input} flags={flags} done={complete}/>);
        function complete() {
            expect(lastFrame()).toMatchSnapshot();
            done();
        }
    });
    xtest('add eslint --use-react', done => {
        const input = ['add', 'eslint'];
        const flags = {useReact: true, skipInstall};
        const {lastFrame} = render(<Tomo input={input} flags={flags} done={complete}/>);
        function complete() {
            expect(lastFrame()).toMatchSnapshot();
            done();
        }
    });
});
