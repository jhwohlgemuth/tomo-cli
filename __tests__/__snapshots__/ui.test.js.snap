// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommandError can render component and create errors file 1`] = `
"{
    name: tomo-test,
    children: [
        {
            name: tomo-errors.txt,
            size: 0,
            type: file
        }
    ],
    size: 0,
    type: directory
}
"
`;

exports[`CommandError can render component and create errors file 2`] = `
"
 [31m[1m✖ [22m[39mSomething has gone horribly [31m[1mwrong[22m[39m
   ↳ [2mDetails written to ./tomo-errors.txt[22m"
`;

exports[`Task component can render (completed) 1`] = `"   [32m[1m✔ [22m[39m[2mtest task text[22m"`;

exports[`Task component can render (errored) 1`] = `"   [31m[1m✖ [22m[39mtest task text"`;

exports[`Task component can render (pending) 1`] = `"   [36m⠋[39m[39m[36m [39m[39m[39mtest task text[39m"`;

exports[`Task component can render (skipped) 1`] = `"   [2m[1m✔ [2m[22m[2mtest task text[22m"`;

exports[`Task component can render with default text 1`] = `"   [32m[1m✔ [22m[39m[2mtask description[22m"`;

exports[`Warning can render 1`] = `
"[33m[39m
[33m ┌─────────┐[39m
 [33m│[39m [33moops...[39m [33m│[39m
[33m └─────────┘[39m
    ↳ Hello World

      [2mPress [22m[1mENTER[22m[2m to continue[22m
"
`;

exports[`populateQueue function can only run tasks that pass condition 1`] = `
Object {
  "foo": "bar",
  "isNotOffline": true,
}
`;

exports[`populateQueue function can only run tasks that pass condition 2`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "online": true,
      },
      "type": "status",
    },
  ],
  Array [
    Object {
      "payload": 0,
      "type": "complete",
    },
  ],
  Array [
    Object {
      "payload": 1,
      "type": "skipped",
    },
  ],
  Array [
    Object {
      "payload": 2,
      "type": "complete",
    },
  ],
  Array [
    Object {
      "payload": 3,
      "type": "skipped",
    },
  ],
]
`;

exports[`populateQueue function can run tasks with options and dispatch task results 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "online": true,
      },
      "type": "status",
    },
  ],
  Array [
    Object {
      "payload": 0,
      "type": "complete",
    },
  ],
  Array [
    Object {
      "payload": 1,
      "type": "complete",
    },
  ],
  Array [
    Object {
      "payload": 2,
      "type": "complete",
    },
  ],
]
`;

exports[`tomo add 1`] = `
"
 [36mAdd automated [1maccessibility[22m testing[39m

 [36m[1m→[22m[39m [39m[36m[1ma11y[22m[39m[39m
   babel
   browsersync
   electron
   esdoc
   eslint
   jest
   makefile
   parcel
   postcss
   rollup
   webpack
"
`;

exports[`tomo add 2`] = `
"
 [36mUse next generation JavaScript, [1mtoday![22m[39m

   a11y
 [36m[1m→[22m[39m [39m[36m[1mbabel[22m[39m[39m
   browsersync
   electron
   esdoc
   eslint
   jest
   makefile
   parcel
   postcss
   rollup
   webpack
"
`;

exports[`tomo add eslint --use-react 1`] = `
"[36m[39m
[36m ╭────────────╮[39m
 [36m│[39m [37m[1madd eslint[22m[39m [36m│[39m
[36m ╰────────────╯[39m
   [32m[1m✔ [22m[39m[2mCreate Babel config file[22m
   [2m[1m✔ [2m[22m[2mInstall Babel core, CLI, presets, and plugins[22m
   [32m[1m✔ [22m[39m[2mAdd React support to Babel configuration file[22m
   [32m[1m✔ [22m[39m[2mCreate ESLint configuration and .eslintignore files[22m
   [2m[1m✔ [2m[22m[2mAdd lint tasks to package.json[22m
   [2m[1m✔ [2m[22m[2mInstall ESLint dependencies[22m
   [36m⠦[39m[39m[36m [39m[39m[39mAdd React support to ESLint configuration file[39m

"
`;

exports[`tomo add eslint 1`] = `
"[36m[39m
[36m ╭────────────╮[39m
 [36m│[39m [37m[1madd eslint[22m[39m [36m│[39m
[36m ╰────────────╯[39m
   [32m[1m✔ [22m[39m[2mCreate Babel config file[22m
   [2m[1m✔ [2m[22m[2mInstall Babel core, CLI, presets, and plugins[22m
   [32m[1m✔ [22m[39m[2mCreate ESLint configuration and .eslintignore files[22m
   [2m[1m✔ [2m[22m[2mAdd lint tasks to package.json[22m
   [2m[1m✔ [2m[22m[2mInstall ESLint dependencies[22m

"
`;
